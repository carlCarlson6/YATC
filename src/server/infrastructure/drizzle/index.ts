import * as baseSchema from "src/server/infrastructure/drizzle/base.drizzle.schema";
import * as followsSchema from "src/server/user/follows/follows.drizzle.schema"
import * as emojisSchema from "src/server/publish-emojeet/emojis.drizzle.schema";
import { sql } from '@vercel/postgres';
//import { drizzle } from 'drizzle-orm/vercel-postgres';
import { drizzle } from 'drizzle-orm/postgres-js';

import postgres from 'postgres';
import { env } from "src/env.mjs";


export const drizzleSchema = {
	...baseSchema,
  ...followsSchema,
  ...emojisSchema,
};

//export const drizzleDb = drizzle(sql, { schema: drizzleSchema } );

export const drizzleDb = drizzle(postgres(env.NEON_DB), { schema: drizzleSchema } );
export type DrizzleDb = typeof drizzleDb;