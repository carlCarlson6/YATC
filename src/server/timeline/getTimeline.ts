import type { Emojeet, EmojiEntityWithReactions } from "./EmojiTweet";

export const getTimeline = ({ 
  fetchEmojeets, addUserData 
}: {
  fetchEmojeets: (userId: string) => Promise<EmojiEntityWithReactions[]>;
  addUserData: (tweets: EmojiEntityWithReactions[]) => Promise<Emojeet[]>;
}) => async (userId: string) => {
  const builtTimeline = await fetchEmojeets(userId);
  return await addUserData(builtTimeline);
};
