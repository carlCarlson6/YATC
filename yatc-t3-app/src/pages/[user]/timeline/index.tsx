import { GetServerSideProps, InferGetServerSidePropsType } from "next";
import { getSession } from "next-auth/react";
import { mockTimeLine } from "../mock-timeline";
import { Box } from "@radix-ui/themes";
import { TimeLineControls } from "./timeline-controls";
import { TimelineDisplay } from "./timeline-display";

export default ({timeline}: TimelineProps) => (
  <Box >
      <TimeLineControls />
      <TimelineDisplay timeline={timeline} />
  </Box>
);

export const getServerSideProps: GetServerSideProps<{timeline: Tweet[]}> =async (context) => {
  const session = await getSession(context)
	if (!session) {
		return {
			redirect: {
				destination: '/',
				permanent: false,
			},
		}
	}

  return { props: {
    timeline: mockTimeLine
  }}
}

type TimelineProps = InferGetServerSidePropsType<typeof getServerSideProps>;

export type Tweet = {
  text: string,
  user: {
    name: string,
    avatar: string,
  } 
}