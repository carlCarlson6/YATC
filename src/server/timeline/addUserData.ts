import { inArray } from "drizzle-orm";
import { type DrizzleDb } from "../infrastructure/drizzle";
import type { EmojiEntityWithReactions, Timeline } from "./EmojiTweet";
import { usersTable } from "../infrastructure/drizzle/base.drizzle.schema";

export const addUserData = (db: DrizzleDb) => async (emojis: EmojiEntityWithReactions[]): Promise<Timeline> => {
  if (emojis.length === 0) return [];

  const usersData = await db.select({
    id: usersTable.id,
    name: usersTable.name,
    avatar: usersTable.image
  }).from(usersTable).where(inArray(usersTable.id, emojis.map(t => t.publishedBy))).execute();

  return emojis.map(emoji => {
    const maybeUser = usersData.filter(x => x.id === emoji.publishedBy).map(x => ({
      id: x.id,
      name: x.name ?? "",
      avatar: x.avatar ?? ""
    })).at(0);
    return {
      ...emoji,
      user: maybeUser ?? { id: "", name: "", avatar: "" }
    };
  });
};
