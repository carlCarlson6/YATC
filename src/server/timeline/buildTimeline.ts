import { desc, eq, inArray } from "drizzle-orm";
import { type DrizzleDb } from "../infrastructure/drizzle";
import { followsTable } from "../user/follows/follows.drizzle.schema";
import type { Timeline } from "./EmojiTweet";
import { type EmojiEntity, emojisTable } from "../emojeets/publish/emojis.drizzle.schema";
import { usersTable } from "../infrastructure/drizzle/base.drizzle.schema";
import { emojisReactionsTable } from "../emojeets/react/emojisReactions.drizzle.schema";

export type EmojiEntityWithReactions = EmojiEntity & {reactions: {emoji: string}[]}

export const buildTimelineFromDb = (db: DrizzleDb) => async (userId: string): Promise<EmojiEntityWithReactions[]> => {
  const queryFollowsEmojisResult = await db
    .select({ emoji: emojisTable })
    .from(followsTable)
    .where(eq(followsTable.userId, userId))
    .innerJoin(emojisTable, eq(followsTable.userId, emojisTable.publishedBy))
    .orderBy(desc(emojisTable.publishedAt))
    .execute();
  const followEmojis = queryFollowsEmojisResult.map(x => ({...x.emoji, reactions: []}));

  const userEmojis = await db
    .select()
    .from(emojisTable)
    .where(eq(emojisTable.publishedBy, userId))
    .orderBy(desc(emojisTable.publishedAt))
    .execute();
  const userEmojisWithReactionsPromises = userEmojis.map(async x => {
    const reactions = await db
      .select({ emoji: emojisReactionsTable.reaction })
      .from(emojisReactionsTable)
      .where(eq(emojisReactionsTable.reactsTo, x.id))
      .execute();
    return {
      ...x,
      reactions
    }
  });
  const userEmojisWithReactions = await Promise.all(userEmojisWithReactionsPromises);

  return [...userEmojisWithReactions, ...followEmojis].sort((a,b) => Number.parseFloat(a.publishedAt) - Number.parseFloat(b.publishedAt)).reverse();
}

export const addUserData = (db: DrizzleDb) => async (emojis: EmojiEntityWithReactions[]): Promise<Timeline> => {
  if (emojis.length === 0) return [];
  
  const usersData = await db.select({
    id: usersTable.id,
    name: usersTable.name,
    avatar: usersTable.image
  }).from(usersTable).where(inArray(usersTable.id, emojis.map(t => t.publishedBy))).execute();

  return emojis.map(emoji => {
    const maybeUser = usersData.filter(x => x.id === emoji.publishedBy).map(x => ({
      id: x.id,
      name: x.name ?? "",
      avatar: x.avatar ?? ""
    })).at(0);
    return {
      ...emoji,
      user: maybeUser ?? { id: "", name: "", avatar: "" }
    };
  });
}