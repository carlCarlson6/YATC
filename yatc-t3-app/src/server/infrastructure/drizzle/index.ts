import { Client } from "@planetscale/database";
import { drizzle } from "drizzle-orm/planetscale-serverless";
import { env } from "yact/env.mjs";
import * as baseSchema from "yact/server/infrastructure/drizzle/base.drizzle.schema";
import * as tweetSchema from "yact/server/send-tweet/tweet.drizzle.schema";
import * as followsSchema from "yact/server/user/follow/follow.drizzle.schema"

const schema = {
	...baseSchema,
  ...tweetSchema,
  ...followsSchema,
};

export const drizzleDb = drizzle(
  new Client({
    url: env.DATABASE_URL,
  }).connection(),
  { schema }
);

export type DrizzleDb = typeof drizzleDb;