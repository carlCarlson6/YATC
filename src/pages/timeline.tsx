import type { GetServerSideProps, InferGetServerSidePropsType } from "next";
import { Box } from "@radix-ui/themes";
import { TimeLineControls } from "../ui/timeline/controls/TimelineControls";
import { TimelineDisplay } from "../ui/timeline/TimelineDisplay";
import { buildTimelineFromDb, addUserData } from "src/server/timeline/buildTimeline";
import { TimelineProvider } from "src/ui/timeline/store";
import { authPageGuard } from "src/server/infrastructure/nextauth/page-auth-guard";
import type { User } from "src/server/user/user";
import Head from "next/head";
import { getTimeline } from "src/server/timeline/getTimeline";
import { vercelKvCache } from "src/server/infrastructure/vercelKv";
import { drizzleDb } from "src/server/infrastructure/drizzle";
import type { Timeline } from "src/server/core/Tweet";

export default function Timeline({serverTimeline, user}: TimelineProps) {
  return (<>
    <Head>
      <title>YATC | TIMELINE</title>
    </Head>
    <TimelineProvider timeline={serverTimeline}>
      <Box>
        <TimeLineControls user={user}/>
        <TimelineDisplay />
      </Box>
    </TimelineProvider>
  </>);
}

export const getServerSideProps: GetServerSideProps<{
  serverTimeline: Timeline,
  user: User
}> = async (context) => {
  const authResult = await authPageGuard(context);
  if (authResult.result == "unauthenticated")
    return authResult.redirectReturn;

  return {
    props: {
      user: authResult.user,
      serverTimeline: await getTimeline({
        cache: vercelKvCache,
        buildTimeline: buildTimelineFromDb(drizzleDb),
        addUserData: addUserData(drizzleDb),
      })(authResult.user.id),
    }
  }
}

type TimelineProps = InferGetServerSidePropsType<typeof getServerSideProps>;