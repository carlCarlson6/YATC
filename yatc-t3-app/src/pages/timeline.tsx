import type { GetServerSideProps, InferGetServerSidePropsType } from "next";
import { Box } from "@radix-ui/themes";
import { TimeLineControls } from "../ui/timeline/controls/TimelineControls";
import { TimelineDisplay } from "../ui/timeline/TimelineDisplay";
import { Timeline, buildTimeline } from "yact/server/timeline/build-timeline";
import { TimelineProvider } from "yact/ui/timeline/store";
import { authPageGuard } from "yact/server/infrastructure/nextauth/page-auth-guard";
import type { User } from "yact/server/user/user";

const Timeline = ({serverTimeline, user}: TimelineProps) => (
  <TimelineProvider timeline={serverTimeline}>
    <Box>
      <TimeLineControls user={user}/>
      <TimelineDisplay />
    </Box>
  </TimelineProvider>
);

export default Timeline;

export const getServerSideProps: GetServerSideProps<{
  serverTimeline: Timeline,
  user: User
}> = async (context) => {
  const authResult = await authPageGuard(context);
  if (authResult.result == "unauthenticated")
    return authResult.redirectReturn;

  return {
    props: {
      user: authResult.user,
      serverTimeline: await buildTimeline(),
    }
  }
}

type TimelineProps = InferGetServerSidePropsType<typeof getServerSideProps>;