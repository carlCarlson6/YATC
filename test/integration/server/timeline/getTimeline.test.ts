import { expect, test } from 'vitest';
import { setupDockerTestDb } from "test/helpers/setupTestingDb";
import { getTimeline } from 'src/server/timeline/getTimeline';
import { buildTimelineFromDb } from 'src/server/timeline/buildTimeline';
import { addUserData } from 'src/server/timeline/addUserData';
import { emojisTable } from 'src/server/emojeets/publish/emojis.drizzle.schema';
import { DrizzleDb } from 'src/server/infrastructure/drizzle';
import { usersTable } from 'src/server/infrastructure/drizzle/base.drizzle.schema';
import { User } from 'src/server/user/profile/userProfile.drizzle.schema';
import { emojisReactionsTable } from 'src/server/emojeets/react/emojisReactions.drizzle.schema';
import { followsTable } from 'src/server/user/follows/follows.drizzle.schema';

const userId = "024a0448-10d8-4a9b-9b0d-0c55caccd4c4";
const anotherUserId = "cd192ae5-39f0-419e-86a7-6c48cc40a7ac";

const withUser = (db: DrizzleDb, user: User) => db.insert(usersTable).values({
  id: user.id,
  name: user.name,
  email: `${user.name}@mail.com`,
  image: user.avatar
}).execute();

const getUserTimeline = async (db: DrizzleDb) => await getTimeline({
  fetchEmojeets: buildTimelineFromDb(db),
  addUserData: addUserData(db)
})(userId);

test("GivenUserWithSomeEmojeets_WhenGetTimeline_ThenTimelineIsReturned", async () => {
  const {db} = await setupDockerTestDb();
  await withUser(db, {id: userId, name: "some-user-name", avatar: "some-user-image"});
  await db.insert(emojisTable).values([
    {
      id:           "277cc7a2-2b75-4303-9179-1538360f8239",
      emoji:        "üêπ",
      publishedAt:  "0",
      publishedBy:  userId
    },
    {
      id:           "bf2c6da7-aa5c-43fd-9a3b-9fd0f2650ec4",
      emoji:        "üê±",
      publishedAt:  "1",
      publishedBy:  userId
    }
  ]).execute();

  const timeline = await getUserTimeline(db);
  expect(timeline).toMatchSnapshot();
});

test("GivenUserWithEmojeetWithReactions_WhenGetTimeline_ThenTimelineWithReactionsIsReturned", async () => {
  const {db} = await setupDockerTestDb();
  await withUser(db, {id: userId, name: "some-user-name", avatar: "some-user-image"});
  await db.insert(emojisTable).values({
    id:           "277cc7a2-2b75-4303-9179-1538360f8239",
    emoji:        "üêπ",
    publishedAt:  "0",
    publishedBy:  userId
  }).execute();
  await db.insert(emojisReactionsTable).values([
    {
      id:           "eea498ae-b8dd-4102-9575-6618f4327f9d",
      reaction:     "üßÄ",
      reactsTo:     "277cc7a2-2b75-4303-9179-1538360f8239",
      publishedAt:  "1",
      publishedBy:  userId
    },
    {
      id:           "2df9f46d-e865-4737-8050-fb5c3e8b0046",
      reaction:     "üçï",
      reactsTo:     "277cc7a2-2b75-4303-9179-1538360f8239",
      publishedAt:  "2",
      publishedBy:  userId
    },
  ]).execute();

  const timeline = await getUserTimeline(db);
  expect(timeline).toMatchSnapshot();
});

test("GivenUserFollowingAnotherUser_WhenGetTimeline_ThenTimelineIsReturned", async () => {
  const {db} = await setupDockerTestDb();

  await withUser(db, {id: userId, name: "some-user-name", avatar: "some-user-image"});
  await withUser(db, {id: anotherUserId, name: "another-user-name", avatar: "another-user-image"});
  await db.insert(followsTable).values({
    id: "1",
    userId,
    isFollowingUserId: anotherUserId
  }).execute();

  await db.insert(emojisTable).values([
    {
      id:           "277cc7a2-2b75-4303-9179-1538360f8239",
      emoji:        "üêπ",
      publishedAt:  "0",
      publishedBy:  userId
    },
    {
      id:           "432b5c9e-de62-4249-b0c0-d452ca3234d6",
      emoji:        "üê±",
      publishedAt:  "0",
      publishedBy:  anotherUserId
    }
  ]).execute();

  const timeline = await getUserTimeline(db);
  expect(timeline).toMatchSnapshot();
});

test("GivenUserFollowingAnotherUser_WithReactions_WhenGetTimeline_ThenTimelineIsReturned", async () => {
  const {db} = await setupDockerTestDb();
  
  await withUser(db, {id: userId, name: "some-user-name", avatar: "some-user-image"});
  await withUser(db, {id: anotherUserId, name: "another-user-name", avatar: "another-user-image"});
  await db.insert(followsTable).values({
    id: "1",
    userId,
    isFollowingUserId: anotherUserId
  }).execute();

  await db.insert(emojisTable).values({
    id:           "277cc7a2-2b75-4303-9179-1538360f8239",
    emoji:        "üêπ",
    publishedAt:  "0",
    publishedBy:  anotherUserId
  }).execute();
  await db.insert(emojisReactionsTable).values({
    id:           "6566e66d-ff4c-451b-a9f4-3ec1393b9aa3",
    reaction:     "ü™§",
    reactsTo:     "277cc7a2-2b75-4303-9179-1538360f8239",
    publishedAt:  "1",
    publishedBy:  anotherUserId
  }).execute();

  const timeline = await getUserTimeline(db);
  expect(timeline).toMatchSnapshot();
});