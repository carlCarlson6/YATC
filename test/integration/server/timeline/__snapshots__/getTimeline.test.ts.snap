// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GivenUserFollowingAnotherUser_WhenGetTimeline_ThenTimelineIsReturned 1`] = `
[
  {
    "emoji": "🐱",
    "id": "432b5c9e-de62-4249-b0c0-d452ca3234d6",
    "publishedAt": "0",
    "publishedBy": "cd192ae5-39f0-419e-86a7-6c48cc40a7ac",
    "reactions": [],
    "user": {
      "avatar": "another-user-image",
      "id": "cd192ae5-39f0-419e-86a7-6c48cc40a7ac",
      "name": "another-user-name",
    },
  },
  {
    "emoji": "🐹",
    "id": "277cc7a2-2b75-4303-9179-1538360f8239",
    "publishedAt": "0",
    "publishedBy": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
    "reactions": [],
    "user": {
      "avatar": "some-user-image",
      "id": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
      "name": "some-user-name",
    },
  },
]
`;

exports[`GivenUserFollowingAnotherUser_WithReactions_WhenGetTimeline_ThenTimelineIsReturned 1`] = `
[
  {
    "emoji": "🐹",
    "id": "277cc7a2-2b75-4303-9179-1538360f8239",
    "publishedAt": "0",
    "publishedBy": "cd192ae5-39f0-419e-86a7-6c48cc40a7ac",
    "reactions": [
      {
        "emoji": "🪤",
      },
    ],
    "user": {
      "avatar": "another-user-image",
      "id": "cd192ae5-39f0-419e-86a7-6c48cc40a7ac",
      "name": "another-user-name",
    },
  },
]
`;

exports[`GivenUserWithEmojeetWithReactions_WhenGetTimeline_ThenTimelineWithReactionsIsReturned 1`] = `
[
  {
    "emoji": "🐹",
    "id": "277cc7a2-2b75-4303-9179-1538360f8239",
    "publishedAt": "0",
    "publishedBy": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
    "reactions": [
      {
        "emoji": "🧀",
      },
      {
        "emoji": "🍕",
      },
    ],
    "user": {
      "avatar": "some-user-image",
      "id": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
      "name": "some-user-name",
    },
  },
]
`;

exports[`GivenUserWithSomeEmojeets_WhenGetTimeline_ThenTimelineIsReturned 1`] = `
[
  {
    "emoji": "🐱",
    "id": "bf2c6da7-aa5c-43fd-9a3b-9fd0f2650ec4",
    "publishedAt": "1",
    "publishedBy": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
    "reactions": [],
    "user": {
      "avatar": "some-user-image",
      "id": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
      "name": "some-user-name",
    },
  },
  {
    "emoji": "🐹",
    "id": "277cc7a2-2b75-4303-9179-1538360f8239",
    "publishedAt": "0",
    "publishedBy": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
    "reactions": [],
    "user": {
      "avatar": "some-user-image",
      "id": "024a0448-10d8-4a9b-9b0d-0c55caccd4c4",
      "name": "some-user-name",
    },
  },
]
`;
